packages/vm/src/imports.rs

```diff
@@ -37,172 +33,222 @@ const MAX_LENGTH_QUERY_CHAIN_REQUEST: usize = 64 * KI;
+pub fn native_db_next<S: Storage, Q: Querier>(env: &Env<S, Q>, iterator_id: u32) -> VmResult<u32> {
+    do_next(env, iterator_id)
+}
+
+//
+// Import implementations
+//
+
 /// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<u32> {
-    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
+fn do_read<S: Storage, Q: Querier>(env: &Env<S, Q>, key_ptr: u32) -> VmResult<u32> {
+    let key = read_region(&env.memory(), key_ptr, MAX_LENGTH_DB_KEY)?;
 
-    let (result, gas_info) =
-        with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.get(&key)))?;
-    process_gas_info::<S, Q>(ctx, gas_info)?;
+    let (result, gas_info) = env.with_storage_from_context::<_, _>(|store| Ok(store.get(&key)))?;
+    process_gas_info::<S, Q>(env, gas_info)?;
     let value = result?;
 
     let out_data = match value {
         Some(data) => data,
         None => return Ok(0),
     };
-    write_to_contract::<S, Q>(ctx, &out_data)
+    write_to_contract::<S, Q>(env, &out_data)
 }
 
 /// Writes a storage entry from Wasm memory into the VM's storage
-pub fn do_write<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
-    key_ptr: u32,
-    value_ptr: u32,
-) -> VmResult<()> {
-    if is_storage_readonly::<S, Q>(ctx) {
+fn do_write<S: Storage, Q: Querier>(env: &Env<S, Q>, key_ptr: u32, value_ptr: u32) -> VmResult<()> {
+    if env.is_storage_readonly() {
         return Err(VmError::write_access_denied());
     }
 
-    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
-    let value = read_region(ctx, value_ptr, MAX_LENGTH_DB_VALUE)?;
+    let key = read_region(&env.memory(), key_ptr, MAX_LENGTH_DB_KEY)?;
+    let value = read_region(&env.memory(), value_ptr, MAX_LENGTH_DB_VALUE)?;
 
     let (result, gas_info) =
-        with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.set(&key, &value)))?;
-    process_gas_info::<S, Q>(ctx, gas_info)?;
+        env.with_storage_from_context::<_, _>(|store| Ok(store.set(&key, &value)))?;
+    process_gas_info::<S, Q>(env, gas_info)?;
     result?;
 
     Ok(())
 }
 
-pub fn do_remove<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<()> {
-    if is_storage_readonly::<S, Q>(ctx) {
+fn do_remove<S: Storage, Q: Querier>(env: &Env<S, Q>, key_ptr: u32) -> VmResult<()> {
+    if env.is_storage_readonly() {
         return Err(VmError::write_access_denied());
     }
 
-    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
+    let key = read_region(&env.memory(), key_ptr, MAX_LENGTH_DB_KEY)?;
 
     let (result, gas_info) =
-        with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.remove(&key)))?;
-    process_gas_info::<S, Q>(ctx, gas_info)?;
+        env.with_storage_from_context::<_, _>(|store| Ok(store.remove(&key)))?;
+    process_gas_info(env, gas_info)?;
     result?;
 
     Ok(())
 }
 
 pub fn do_canonicalize_address<A: Api, S: Storage, Q: Querier>(
```

Comment:
Okay, so we don't need to use the native_* format for all callbacks, but can also use these that capture Api?
What is the cost of this/benefit of native functions?

If there is a clear benefit to native functions (beyond readability), we could consider to store the `Api` in the `Env` as well, and this could be native function as well

Reply:
That's a pretty good idea, as copying env is cheap and we don't need to create new functions for every instance. It just adds another generic argument to Env.